### ISSUE TITLE
Locked files seems to cause corrupt backup

### ISSUE THREAD

**Issue by balchen** [duplicati/duplicati#3594]
- [X] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.4.5_beta_2018-11-28
- **Operating system**: Windows Server 2012 R2
- **Backend**: IIS FTP

## Description

I've configured a backup of our Atlassian Stash data. The backup failed every time with "One or more errors occurred". I clicked "Show log", but there were no log entries for "General", and "Remote" showed only successful uploads of data files. Re-running the backup without changes frequently caused error messages "Detected non-empty blockset with no associated blocks". Deleting and recreating the database failed. The only way to continue was to delete the database, delete the remote files, and start over.

Since the GUI log showed no entries, I configured file logging and was able to identify a possible reason for the issues. It all seems to start when Duplicati hits a file that Stash has locked (there are apparently several of them), which causes this:

```
2019-01-08 23:38:35 +01 - [Warning-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-PathProcessingFailed]: Failed to process path: C:\Atlassian\ApplicationData\Stash\plugins\.osgi-cache\felix\felix-cache\cache.lock
System.IO.IOException: The process cannot access the file because another process has locked a portion of the file.
[stacktrace]
2019-01-08 23:50:37 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-b444ad11794f74db2a057e860451bc346.dblock.zip (499.90 MB)
2019-01-08 23:50:37 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-i2b73e5165b7a41de8c5ad97a34807d47.dindex.zip (410.63 KB)
2019-01-08 23:50:39 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-i2b73e5165b7a41de8c5ad97a34807d47.dindex.zip (410.63 KB)
2019-01-08 23:50:39 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b4def9402bd69477fb47bcd889109b69d.dblock.zip (499.99 MB)
2019-01-08 23:50:41 +01 - [Error-Duplicati.Library.Main.Operation.BackupHandler-FatalError]: Fatal error
System.Exception: Detected non-empty blocksets with no associated blocks!
[stacktrace]
```

and the backup terminates. The suggested resolution for non-empty blocksets with no associated blocks seems to be to repair or recover the local DB, so I did that:

```
2019-01-09 00:01:25 +01 - [Information-Duplicati.Library.Main.Controller-StartingOperation]: The operation Repair has started
2019-01-09 00:01:25 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Started:  ()
2019-01-09 00:01:26 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Completed:  (15 bytes)
2019-01-09 00:01:26 +01 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Temporary: duplicati-20190108T222340Z.dlist.zip
2019-01-09 00:01:26 +01 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-Remove incomplete file]: removing incomplete remote file listed as Uploading: duplicati-b4def9402bd69477fb47bcd889109b69d.dblock.zip
2019-01-09 00:01:26 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Delete - Started: duplicati-b4def9402bd69477fb47bcd889109b69d.dblock.zip ()
2019-01-09 00:01:27 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Delete - Completed: duplicati-b4def9402bd69477fb47bcd889109b69d.dblock.zip ()
2019-01-09 00:01:29 +01 - [Information-Duplicati.Library.Main.Operation.FilelistProcessor-RemoteUnwantedMissingFile]: removing file listed as Temporary: duplicati-bd215a83de70d4b49895925c5aa2737c0.dblock.zip
2019-01-09 00:01:30 +01 - [Information-Duplicati.Library.Main.Operation.RepairHandler-DatabaseIsSynchronized]: Destination and database are synchronized, not making any changes
2019-01-09 00:02:07 +01 - [Information-Duplicati.Library.Main.Controller-StartingOperation]: The operation Backup has started
2019-01-09 00:02:23 +01 - [Error-Duplicati.Library.Main.Operation.BackupHandler-FatalError]: Fatal error
System.Exception: Detected non-empty blocksets with no associated blocks!
[stacktrace]
```

Obviously something was in an invalid state from the terminated backup, which the repair job tried to fix, but it did not resolve the issue. To recover from this issue, I had to delete the local database, delete all files in the remote location, make sure that all potentially locked files were excluded/enable shadow copy, and restart from scratch.

## Steps to reproduce

1. Configure and run a backup that will hit locked files.
2. Attempt to recover from corrupt state by running database repair, with no success

- **Actual result**:
Locked files cause corrupt backup that cannot be recovered from
- **Expected result**:
Backup must not be corrupted by locked files. Correct way to resolve locked files is debatable, but resolution must never cause corrupt backup.

**Comment by balchen** [comment]
The error message says "has locked a portion of the file". I can see if I can write some code that does byte range locking and do a test with a different Duplicati setup to confirm.

**Comment by balchen** [comment]
I wrote a simple console app to lock a file in every way possible (in .NET) and configured Duplicate to back up the files. All the different locks cause Duplicati to log an IOException (either the file is completely locked or parts are locked), but nothing makes the backup fail. So locking might not be the culprit after all, and it's just a coincidence that the failures stopped when I excluded the locked files.

I can provide the complete log file from the failed and successful backups (info log level) if that helps. I'll also start looking at other causes.

**Comment by balchen** [comment]
When I was deleting and recreating the backup job to start over after database corruption, and was altering the configuration to exclude locked files, I see that I at one point changed the FTP backup from FTP (Alternate) to FTP. I chose FTP (Alternate) in the first place because it was able to report success on "Test connection"; FTP was never able to complete a test. But this seems to have other effects as well. FTP (Alternate) is much worse at connection handling (e.g. #3592) and "Channel is retired" exceptions are frequent with FTP (Alternate) and not reported at all with FTP. 

The coincidence might be that I both excluded locked files and change the backend to FTP, and that in reality it was the change to FTP that made the issues disappear.

**Comment by Phaiax** [comment]
I also have this error one one of the machines and tried to investigate.

While I will not share the whole .sqlite database, I'll share the parts that I think contain the error.

The SQL query that detects the 'non-empty blocksets with no associated blocks' gives the following result (`COUNT(*)` replaced by `*`):
```
SELECT * FROM Blockset WHERE Length > 0 AND ID NOT IN (SELECT BlocksetId FROM BlocksetEntry)
=> ID="51339"	LENGHT="6646562"	HASH="LRFelwMN6lAdLHz+NSK6yNqz/vW5e7l2WfibCVFoQXQ="
```

There is no associated entry in FileLookup with that blockid:
```
SELECT COUNT(*) FROM FileLookup where ID = 51339
=> 0
```

I [searched](https://github.com/Phaiax/accesstest) for a file in the backupfolder with that exact file size and found
```
 SIZE=6646562 | C:\Stefan\Bilder\2017\Ochsenplatz\P1360586.JPG
```

Then I looked in the `FileLookup` table for the other files from that folder.

[Here](https://imgur.com/QQXOOaA) is a screenshot of the area from sqlitebrowser.

You'll notice strange things:
 - The folder `Neujahrsempfang` is inserted twice (no other folder is inserted twice as far as I have skimmed)
 - The missing file would fit exactly where the first Neujahrsempfang folder is inserted
 - The ID jumps by 17 before the first 'Neujahrsempfang'

I also verified further:
 - This is the only jump in the ID field in the whole table
 - All files belong to the same "RemoteOperation". (The first and the last blocks of the Blocks of the surrounding files all belong to RemoteVolumes that have OperationID=2
 - Anyway, even if there are 80 Operations in the Operation table, all except the first Remotevolume belong to OperationID=2. 
 - There are no other files missing in the folder `Neujahrsempfang` and the folder below and above.

Futher discoveries which may not be a problem at all:
 - During the middle of the operation, there were Remotevolumes deleted (see [table Remotevolume as csv on pastebin](https://pastebin.com/raw/YdiMFpNW)
 - For MetadataID=27313 (the first Neujahrsempfang folder) there is no entry in the Metadata table
 - For MetadataID=27330 (the second Neujahrsempfang folder), there is a metadata entry and the metadata in the block (manually decrypted) looks valid: `{"win-ext:accessrules":"[]","CoreAttributes":"Directory","CoreLastWritetime":"636233899732360333","CoreCreatetime":"636233894585012421"}`

I have two guesses what could be the problem:
 - Synchronization issues with the sqlite driver (async -> sync adapter)
 - Another (cleanup?) job that inferes with long running backup processes
 - Maybe duplicati starts the backup job again after 24 h?

**Comment by ts678** [comment]
I'm hugely pleased that someone who can actually hit the problem is chasing it into a deeper level. Thanks!

Possibly those clues will mean something to the right experts, but for me some clues on how it breaks from an operation viewpoint would help more. Still fishing for steps to reproduce, then maybe others can help. As an example, I mentioned compact earlier, and I wonder if OperationID=2 might have been one? One can easily look in ordinary backup logs to see if a compact occurs just before the problem occurs, but I haven't heard feedback on that. It does seem odd that a single operation would dominate the table. What was it?

One useful technique for tracing what went on (then went wrong) is to keep a log at the (extremely wordy) --log-file-log-level=profiling with --profile-all-database-queries=true, then you may see exactly when a row got inserted, and maybe even when it got deleted. Sometimes, though, changes are driven by other SELECT results. This is harder to study, but at least one can get candidates by seeing it work on the interesting table.

Another technique is to keep recent history of the database somehow so you can see what it used to look like before it broke. For my work in ["Unexpected difference in fileset" test case and code clue #3800](https://github.com/duplicati/duplicati/issues/3800) I just used a --run-script-before such as the (very crude) .bat file below (expect warnings until files are present):

~~~
SET DB=90807567746884908385
MOVE /Y C:\ProgramData\Duplicati\%DB%.4.sqlite C:\ProgramData\Duplicati\%DB%.5.sqlite
MOVE /Y C:\ProgramData\Duplicati\%DB%.3.sqlite C:\ProgramData\Duplicati\%DB%.4.sqlite
MOVE /Y C:\ProgramData\Duplicati\%DB%.2.sqlite C:\ProgramData\Duplicati\%DB%.3.sqlite
MOVE /Y C:\ProgramData\Duplicati\%DB%.1.sqlite C:\ProgramData\Duplicati\%DB%.2.sqlite
COPY /Y C:\ProgramData\Duplicati\%DB%.sqlite C:\ProgramData\Duplicati\%DB%.1.sqlite
~~~

Great work so far even starting to dig. Are you able to get the error often? How might you get it more? Alternatively, how might you get it less, e.g. if compact is suspected then test --no-auto-compact=true.

Another option is to post a [DB bug report](https://duplicati.readthedocs.io/en/latest/03-using-the-graphical-user-interface/#creating-a-bug-report), which attempts to sanitize pathnames, however I found you should probably also run the SQLite vacuum command if you want to remove freelist holding leftovers.

**Comment by Phaiax** [comment]
Hi

I looked again and found out that, while almost all RemoteVolumes belong to OperationID=2, there are actually a lot of put operations in the RemoteOperation table with OperationID=1. So it is not the first _real_ backup that went wrong, as I thought first.

The OperationID=1 started on 17.2. 19:47 (according to Operations table)
The `put` operations from OperationID=1 started on 17.2. 19:49
and went continually until 18.2. 13:20, which is the last RemoteOperation with OperationID=1

OperationID=2 started on 18.2. 18:02 (according to Operations table)
with a `list` operation at 18.2. 18:12
In the returned list, there was not yet a `dlist` file (in JSON from the `Data` column.) Neither was a `dlist` file uploaded in the RemoteOperations. Is that usual behaviour, that the backup pauses after a certain time?

Anyway, I think the problem may lay somewhere else entierly, so I started a backup job with the log level set to profiling and the bat file to copy the database away

My .bat looks like this:
```
SET DB=123456789

COPY /Y C:\Windows\System32\config\systemprofile\Appdata\Local\Duplicati\%DB%.sqlite C:\Windows\System32\config\systemprofile\Appdata\Local\Duplicati\%DB%.%date:~-4,4%%date:~-7,2%%date:~-10,2%_%time:~0,2%%time:~3,2%%time:~6,2%_%time:~9,2%.sqlite
COPY /Y C:\Windows\System32\config\systemprofile\Appdata\Local\Duplicati\%DB%.sqlite-journal C:\Windows\System32\config\systemprofile\Appdata\Local\Duplicati\%DB%.%date:~-4,4%%date:~-7,2%%date:~-10,2%_%time:~0,2%%time:~3,2%%time:~6,2%_%time:~9,2%.sqlite-journal
```

Is it usual, that the database lives in the Appdata folder of the SYSTEM user? That was the default path.

For the question of how often I get that error: Only on one device, but there it failed every time of the 3 or 4 times I tried (with different Duplicati versions). Last time I tried was Februrary, and I hoped it would be fixed by now^^.

The thing is, I only switched to Duplicati, because the Windows Backup was never successful, too.

The log file will become big.. 180 MB after 45 min...  😄

**Comment by Phaiax** [comment]
A the question I forgot to ask: Is the logfile appended to or overwritten if the next backup starts?

**Comment by ts678** [comment]
> Neither was a `dlist` file uploaded in the RemoteOperations. Is that usual behaviour, that the backup pauses after a certain time?

I'm used to seeing the dlist go up at the end of the backup. Without a dlist on the remote, you can't recreate the database or run a direct restore. Your RemoteVolume table seems to say no dlists are there, however you could look on the remote to see if that's truly the situation or if the DB is misreporting.

> log level set to profiling

and --profile-all-database-queries=true I hope, otherwise some of the high-volume interesting ones don't show. On Linux, you might use `less` to look around. On Windows, [glogg](https://glogg.bonnefon.org/) handles big files (such as these) well, if you just let it grow. It's appended. Another option to keep size to a more reasonable level would be to rotate the logs off as well, but sometimes one big log is easier to work with, e.g. if looking for a specific INSERT by looking for the inserted values, you want a long history. Then again, there are search tools for multiple files... I'm in Notepad++ a lot trying to find my way through Duplicati source files (for lack of any better tools).

**Comment by klorinczi** [comment]
I also get "Detected non-empty blocksets with no associated blocks!" error.
Using Windows 7, Duplicati - 2.0.4.22_canary_2019-06-30.
Remote is Debian 8 SFTP.

**Comment by DisplayNerd** [comment]
So do I. UNRAID system using docker - 2.0.4.3 beta. I tried 3 times

**Comment by AntiBlueQuirk** [comment]
Yep. I think I'm getting this too. The initial backup worked okay (though with lots of warnings about locked files). After trying to update the backup, I got the "Detected non-empty blockset with no associated blocks" before even scanning files. Tried a repair, which seemed to fix the problem, except now while it's scanning/copying files I get "Unexpected result count: 0, expected 1". There's about 10 of these hard errors. Interestingly, one of the files that is causing problems is Duplicati related, though not all of them are. They all seem to be locked though.

I'm on Windows 10 - 2.0.4.22_canary_2019-06-30. I'm backing up to a local drive right now, while I try to make sure I can get a good initial backup. (Which has proven difficult.)

```
2019-09-01 07:54:46 -05 - [Warning-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-PathProcessingFailed]: Failed to process path: C:\Users\Username\AppData\Local\Duplicati\OPZAJHRZBC.sqlite
System.IO.IOException: The process cannot access the file because another process has locked a portion of the file.

   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.ReadCore(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.FileStream.Read(Byte[] array, Int32 offset, Int32 count)
   at System.IO.Stream.<>c.<BeginReadInternal>b__39_0(Object )
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.IO.Stream.EndRead(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncTrimPromise`1.Complete(TInstance thisRef, Func`3 endMethod, IAsyncResult asyncResult, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Utility.Utility.<ForceStreamReadAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.StreamBlockSplitter.<>c__DisplayClass2_0.<<Run>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.StreamBlock.<ProcessStream>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.<>c__DisplayClass1_0.<<Run>b__0>d.MoveNext()
2019-09-01 07:54:46 -05 - [Error-Duplicati.Library.Main.Database.LocalBackupDatabase-CheckingErrorsForIssue1400]: Checking errors, related to #1400. Unexpected result count: 0, expected 1, hash: X2h7M6TnfMNPQQJwKT9gF3QToYk3v7lf09/EbNAkMg0=, size: 102400, blocksetid: 2303979, ix: 20, fullhash: /9gWHHIuOgwoef74yyfw/rO7cd0u87LY6v6n4XFCmfM=, fullsize: 3251608
2019-09-01 07:54:46 -05 - [Error-Duplicati.Library.Main.Database.LocalBackupDatabase-FoundIssue1400Error]: Found block with ID 8789176 and hash X2h7M6TnfMNPQQJwKT9gF3QToYk3v7lf09/EbNAkMg0= and size 83216
2019-09-01 07:54:46 -05 - [Warning-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-PathProcessingFailed]: Failed to process path: C:\Users\Username\AppData\Local\Duplicati\OPZAJHRZBC.sqlite-journal
System.Exception: Unexpected result count: 0, expected 1, check log for more messages
   at Duplicati.Library.Main.Database.LocalBackupDatabase.AddBlockset(String filehash, Int64 size, Int32 blocksize, IEnumerable`1 hashes, IEnumerable`1 blocklistHashes, Int64& blocksetid, IDbTransaction transaction)
   at Duplicati.Library.Main.Operation.Backup.BackupDatabase.<>c__DisplayClass13_0.<AddBlocksetAsync>b__0()
   at Duplicati.Library.Main.Operation.Common.SingleRunner.<>c__DisplayClass4_0`1.<RunOnMain>b__0()
   at Duplicati.Library.Main.Operation.Common.SingleRunner.<DoRunOnMain>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.StreamBlockSplitter.<>c__DisplayClass2_0.<<Run>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.StreamBlock.<ProcessStream>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.<>c__DisplayClass1_0.<<Run>b__0>d.MoveNext()
```

**Comment by ts678** [comment]
[CheckingErrorsForIssue1400 and FoundIssue1400Error test case, analysis, and proposal #3868](https://github.com/duplicati/duplicati/issues/3868) led to at least a partial fix from [Fix for issue1400 #3872](https://github.com/duplicati/duplicati/pull/3872), relating to files growing during backup. The fix is now in canary, if anyone would like to try it earlier on a bleeding-edge release. Not certain it helps locked files case, however that might be a red herring. The only actual error detail that I find here decodes like this:

> 2019-09-01 07:54:46 -05 - [Error-Duplicati.Library.Main.Database.LocalBackupDatabase-CheckingErrorsForIssue1400]: Checking errors, related to #1400. Unexpected result count: 0, expected 1, hash: X2h7M6TnfMNPQQJwKT9gF3QToYk3v7lf09/EbNAkMg0=, size: 102400, blocksetid: 2303979, ix: 20, fullhash: /9gWHHIuOgwoef74yyfw/rO7cd0u87LY6v6n4XFCmfM=, fullsize: 3251608
(Here it's looking for a block via block hash and size which is assumed to be full default 100KB size, but there is none found. 1 was expected, but it only got 0)
> 
> 2019-09-01 07:54:46 -05 - [Error-Duplicati.Library.Main.Database.LocalBackupDatabase-FoundIssue1400Error]: Found block with ID 8789176 and hash X2h7M6TnfMNPQQJwKT9gF3QToYk3v7lf09/EbNAkMg0= and size 83216
(Here is one with the right hash but a shorter size. Bug analysis says this can happen if the file continues to grow after an intermediate end-of-file is reached)
>
> 2019-09-01 07:54:46 -05 - [Warning-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-PathProcessingFailed]: Failed to process path: C:\Users\Username\AppData\Local\Duplicati\OPZAJHRZBC.sqlite-journal
(sqlite journal is a [temporary file](https://www.sqlite.org/tempfiles.html) that probably grows as it goes, so that fits idea)

"Detected non-empty blocksets with no associated blocks!" is mentioned in the cited analysis, and can happen if the file grows slowly enough that the very first block is short, thus not findable or inserted...

**Comment by BlueBlock** [comment]
> (sqlite journal is a temporary file that probably grows as it goes, so that fits idea)

Under normal operation the sqlite journal file is created and removed often. Attempting to backup the journal file will likely always report a problem especially if the file is removed while attempting to back it up.

**Comment by DisplayNerd** [comment]
Wow. Good job. This has been plaguing me for the longest time. I could never complete a backup. Is this a comprehensive fix or will more testing need to be done to check if the bug is squashed for good?

**Comment by BlueBlock** [comment]
The fix for 1400 resolves the problem when a file is growing but not one that is shrinking or deleted.

To me the fundamental issue that needs to be decided is how Duplicati should handle shrinking and deleted files that are in the process of being backed up; should the partial file data be stored in the backup or should the file be removed from the backup?

@ts678 and I spoke a bit about this already but I don't think we came to a resolution.

The benefit of backing up a partial file is that at least you have some of the file.

The downside is that depending on the type of file a partial file might be completely unusable and for a restore the user has no idea if a file is partial or complete. 

I don't like the idea that there is uncertainty and a lack of trust that -all- files contained in a backup are complete files. 

We either include only complete files or have a way to identify files that are partial. This could be done by appending ".partial" to any partial file within the backup. Then a partial file would not overwrite some previous complete file.

Input from others like @kenkendk, @warwickmm and @drwtsn32x would be helpful.

**Comment by drwtsn32x** [comment]
> I don't like the idea that there is uncertainty and a lack of trust that -all- files contained in a backup are complete files.

Snapshots are required to ensure you get a good backup of all files (even in-use files - although they are not necessarily application-consistent).  Snapshots also avoid the entire problem of files growing/shrinking during backup since it's the actual snapshot being backed up, not the live files.

But not everyone can use snapshots and they don't work in every situation.

If Duplicati detects a file is being modified during backup, I think it should back up what it can BUT I think it's critical that a warning alert be triggered.  The job results can show the file(s) that were modified during backup and may not be completely valid.  This way the end user has awareness.

**Comment by BlueBlock** [comment]
> If Duplicati detects a file is being modified during backup, I think it should back up what it can BUT I think it's critical that a warning alert be triggered. The job results can show the file(s) that were modified during backup and may not be completely valid. This way the end user has awareness.

Agreed, snapshot use avoids the problem.

My concern is that warnings in the logs is not enough or easy enough to work with in such a scenario. Consider a scenario with a backup having dozens or hundreds of partial files. For a full restore you would need to walk back through each previous backup to see if it has a complete file. Now do that dozens or hundreds of times.

**Comment by drwtsn32x** [comment]
> Consider a scenario with a backup having dozens or hundreds of partial files.

Do we think this problem is that widespread?  Even if it is, what is the alternative to warning that the job wasn't 100% successful and listing the "problem" files?

**Comment by BlueBlock** [comment]
> Do we think this problem is that widespread? Even if it is, what is the alternative to warning that the job wasn't 100% successful and listing the "problem" files?

Wide enough that issue 1400 is a thing. 

It looks like a lot of other software handles this by simply saying, use VSS, otherwise your backup experience won't be great.

So I guess I'm on the fence as to handling this. So I guess at minimum we can expand the existing file handling for shrinking and deleted files to backup whatever data we can read and log a warning. That should then resolve exceptions from file handling without VSS.

**Comment by DisplayNerd** [comment]
> > Do we think this problem is that widespread? Even if it is, what is the alternative to warning that the job wasn't 100% successful and listing the "problem" files?
> 
> Wide enough that issue 1400 is a thing.
> 
> It looks like a lot of other software handles this by simply saying, use VSS, otherwise your backup experience won't be great.
> 
> So I guess I'm on the fence as to handling this. So I guess at minimum we can expand the existing file handling for shrinking and deleted files to backup whatever data we can read and log a warning. That should then resolve exceptions from file handling without VSS.

Yeah. I agree. My server does not seem to support LVM (or has no documentation). Having a warning if my file wasn't completely backed up would help a lot. Maybe you can add it as an option for the user to determine what kind of warning you would get. Also, issue 1400 was definitely a problem. I found so many users experiencing the same thing. In fact, I don't think I completed a single successful backup myself after 2.0.3.6 was released. Thank you for fixing it. Please don't forget about us non-VSS users!

**Comment by AntiBlueQuirk** [comment]
Yeah, I think it's a pretty important issue. Windows users will hit it automatically if they try to back up their User folders, and don't specifically exclude `USERFOLDER\AppData\Local\Duplicati`. There are probably other active files in the `Local` directory as well. The backup aborts, and the database gets at least somewhat corrupt, which can be really frustrating if you've already spent a long time working on your backup. I should note that I'm a new user, and having my first backup fail because I tried to backup my user folder is not a great experience. (I'm not asking for condolences, just noting that other first-time users might switch to some other program rather than try to sort out the issue.)

I think a warning is definitely in order, possibly with a link to a new entry in the FAQ that explains that you can't really properly back up actively changing files. There really isn't a solution except waiting for the file to close. (Activating snapshots will get you access to locked files, but it won't fix files being in inconsistent states.) There could be a flag to turn the warning into an error for users that need it, much like how you can make snapshots mandatory.

Waiting on locked/changing/open files may actually be a viable option, at least for some files. Duplicati could push it into a queue for later and move onto other files. It could revisit this queue periodically, and finally try to empty it when there are no files left. If it can't finish backing up these files within a reasonable period, it finally gives up on them and issues a warning. This would give a best-effort for the case where a file just happens to be open for writing when the backup starts, For files that are perpetually open/locked, there's really nothing to do. (Of course, this would require a lot of work.)

**Comment by drwtsn32x** [comment]
> My server does not seem to support LVM

LVM has to be set up first - before you create partitions and format filesystems. Unfortunately it's not just a service that enhances existing partitions (like VSS).  But LVM has a lot more capabilities than VSS.

> Having a warning if my file wasn't completely backed up would help a lot.

I think it's an absolute must!

**Comment by BlueBlock** [comment]
From what I've read so far, without VSS, most backup software will lock a file while backing it up.

**Comment by fresheneesz** [comment]
I see this warning every single time it tries to backup:

```
Warning-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-PathProcessingFailed]: Failed to process path: D:\<...>\webrtc\server.key
```

I tried inspecting the properties of the file (in windows 10) and when I go to the Security tab, it tells me "The permissions on server.key are incorrectly ordered, which may cause some entries to be ineffective." Is there any reason that might be related to this issue?

I created the file using the following command (again on windows 10):

```
openssl req -x509 -nodes -days 999 -newkey rsa:2048 -keyout server.key -out server.crt -config req.conf
```

Where req.conf contains (some info changed):

```
[req]
distinguished_name = req_distinguished_name
x509_extensions = v3_req
prompt = no
[req_distinguished_name]
C = US
ST = CA
L = New  York
O = N/A
OU = N/A
CN = 111.111.111.111
[v3_req]
subjectAltName = @alt_names
[alt_names]
DNS.0 = localhost
IP.0 = 111.111.111.111
IP.1 = 111.111.111.112
```

**Comment by drwtsn32x** [comment]
> Is there any reason that might be related to this issue?

It could... as Windows says, some entries of the ACL may be ineffective.  So maybe the Duplicati process doesn't have permission to read the file.

It should be easy to reset the ACL to match the parent folder and fix this issue (assuming the parent folder ACL is also not ordered incorrectly).  If you want to do that, right click on the parent folder, go to Properties, Security, Advanced, and check the box that says "replace all child object permission entries..."

But be careful - this server.key file contains the private key for your certificate. You should protect it!  In fact I wonder if openssl tried to protect it by tweaking the ACL (incorrectly).  I have never used openssl on Windows so I'm not sure.

**Comment by fresheneesz** [comment]
@drwtsn32x Thanks for the info! It wouldn't let me replace the ACL, telling me I didn't have permission or something. I did seem to have permission to delete it tho, so I just copied the contents and deleted the original file. Thanks for the warning, but its just for a test project so it doesn't actually need to be secure or anything ; ) Hopefully this solves the issue!

**Comment by AyrA** [comment]
I have the same issue for partially locked files. In my case, the lockfile from the tor windows client causes a problem. I tried adding the file to the exclusions but it looks like duplicati still tries to access it.
